### PR TITLE
Attribute for configuring composite primary keys

### LINKED ISSUE #11003: KeyAttribute support for composite primary key
I'd like to declare a primary key that consists of multiple columns only with the `Key` attribute, just like for all other keys and as it was supported in EF6. This is the only situation when I have to go to the `OnModelCreating` method in another file to find out about the basic structure of the entity.

Currently I have to write this scattered code outside of where the entity is defined:

	protected override void OnModelCreating(ModelBuilder modelBuilder)
	{
		base.OnModelCreating(modelBuilder);
		modelBuilder.Entity<MyEntity>()
			.HasKey(e => new { e.KeyColumn1, e.KeyColumn2 });
	}

I'd prefer to keep it simple and in a single place, the entity class:

	class MyEntity
	{
		[Key, Column(Order = 0)]
		public int KeyColumn1 { get; set; }

		[Key, Column(Order = 1)]
		public int KeyColumn2 { get; set; }
	}

I could imagine it would be helpful to get rid of the slightly unintuitive `Column` attribute just for this and define the order where it belongs and is used, the `Key` attribute, like in `[Key(Order = 1)]`.

I'm wondering why there isn't already an issue for this. I have found one or two issues that might be similar but I don't understand them well enough to consider them a match for this topic. Also I couldn't see this mentioned in the sparse information I could find about the upcoming EF Core 2.1 so I have to assume it's not included in the next version.

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #11003

This PR introduces a new `[PrimaryKey]` which follows the same pattern as `[Index]` in that it is applied to the entity type class and takes an ordered list of property names. It takes precedence over any `[Key]` attributes on properties, since these may still be needed for OData or other technologies. `PrimaryKey` and `Keyless` cannot be used on the same type.

### CODE REVIEW

File: src/EFCore.Abstractions/PrimaryKeyAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+/// <summary>
+///     Specifies a primary key for the entity type mapped to this CLR type. This attribute can be used for both keys made up of a
+///     single property, and for composite keys made up of multiple properties. `System.ComponentModel.DataAnnotations.KeyAttribute`
+///     can be used instead for single-property keys, in which case the behavior is identical. If both attributes are used, then
+///     this attribute takes precedence.
+/// </summary>
+/// <remarks>
+///     See <see href="https://aka.ms/efcore-docs-modeling">Modeling entity types and relationships</see> for more information and examples.
+/// </remarks>
```
Comment (roji) [inline src/EFCore.Abstractions/PrimaryKeyAttribute.cs:16]: I tend to try to keep the summary short

```suggestion
/// <summary>
///     Specifies a primary key for the entity type mapped to this CLR type.
/// </summary>
/// <remarks>
///     <para>
///         This attribute can be used both for keys made up of a single property, and for composite keys made up of multiple properties.
///         <see cref="KeyAttribute" /> can be used instead for single-property keys, in which case the behavior is identical. If both
///         attributes are used, then this attribute takes precedence.
///     </para>
///     <para>
///         See <see href="https://aka.ms/efcore-docs-modeling">Modeling entity types and relationships</see> for more information and
///         examples.
///     </para>
/// </remarks>
```

File: src/EFCore.Abstractions/PrimaryKeyAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    public PrimaryKeyAttribute(params string[] propertyNames)
+    {
+        Check.NotEmpty(propertyNames, nameof(propertyNames));
+        Check.HasNoEmptyElements(propertyNames, nameof(propertyNames));
+
+        PropertyNames = propertyNames.ToList();
+    }
+
+    /// <summary>
+    ///     The properties which constitute the index, in order.
```
Comment (roji) [inline src/EFCore.Abstractions/PrimaryKeyAttribute.cs:33]:
```suggestion
    ///     The properties which constitute the primary key, in order.
```

File: src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs [hunk 3]
```diff
@@ -42,8 +62,7 @@ public KeyAttributeConvention(ProviderConventionSetBuilderDependencies dependenc
             switch (entityType.GetIsKeylessConfigurationSource())
             {
                 case ConfigurationSource.DataAnnotation:
-                    Dependencies.Logger
-                        .ConflictingKeylessAndKeyAttributesWarning(propertyBuilder.Metadata);
+                    Dependencies.Logger.ConflictingKeylessAndKeyAttributesWarning(propertyBuilder.Metadata);
```
Comment (roji) [inline src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs:65]: Should this block (check for keyless) also be done for ProcessEntityTypeAdded/ProcessEntityTypeBaseTypeChanged?

File: src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs [hunk 4]
```diff
@@ -52,39 +71,49 @@ public KeyAttributeConvention(ProviderConventionSetBuilderDependencies dependenc
...
-            // TODO: Log a warning that KeyAttribute is being ignored. See issue#20014
-            // This code path will also be hit when entity is marked as Keyless explicitly
-            return;
+            return false;
         }
 
-        var entityTypeBuilder = entityType.Builder;
-        var currentKey = entityTypeBuilder.Metadata.FindPrimaryKey();
-        var properties = new List<string> { propertyBuilder.Metadata.Name };
+        var primaryKeyAttributeExists = CheckPrimaryKeyAttributesAndEnsurePrimaryKey(entityType, shouldThrow);
```
Comment (roji) [inline src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs:90]:
```suggestion
        var primaryKeyAttributeExists = CheckPrimaryKeyAttributeAndEnsurePrimaryKey(entityType, shouldThrow);
```

File: src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs [hunk 5]
```diff
@@ -23,13 +28,28 @@ public KeyAttributeConvention(ProviderConventionSetBuilderDependencies dependenc
...
+        IConventionEntityType? newBaseType,
+        IConventionEntityType? oldBaseType,
+        IConventionContext<IConventionEntityType> context)
+    {
+        if (oldBaseType == null)
+        {
+            return;
+        }
+
+        CheckAttributesAndEnsurePrimaryKey((EntityType)entityTypeBuilder.Metadata, null, false);
```
Comment (roji) [inline src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs:49]:
```suggestion
        CheckAttributesAndEnsurePrimaryKey((EntityType)entityTypeBuilder.Metadata, null, shouldThrow: false);
```

File: src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs [hunk 6]
```diff
@@ -23,13 +28,28 @@ public KeyAttributeConvention(ProviderConventionSetBuilderDependencies dependenc
...
-    /// </summary>
-    /// <param name="propertyBuilder">The builder for the property.</param>
-    /// <param name="attribute">The attribute.</param>
-    /// <param name="clrMember">The member that has the attribute.</param>
-    /// <param name="context">Additional information associated with convention execution.</param>
+    /// <inheritdoc />
+    public virtual void ProcessEntityTypeAdded(
+        IConventionEntityTypeBuilder entityTypeBuilder,
+        IConventionContext<IConventionEntityTypeBuilder> context)
+        => CheckAttributesAndEnsurePrimaryKey((EntityType)entityTypeBuilder.Metadata, null, false);
```
Comment (roji) [inline src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs:35]:
```suggestion
        => CheckAttributesAndEnsurePrimaryKey((EntityType)entityTypeBuilder.Metadata, null, shouldThrow: false);
```

File: src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs [hunk 7]
```diff
@@ -95,17 +124,29 @@ public KeyAttributeConvention(ProviderConventionSetBuilderDependencies dependenc
         var entityTypes = modelBuilder.Metadata.GetEntityTypes();
         foreach (var entityType in entityTypes)
         {
+            var primaryKeyAttributeExits = CheckAttributesAndEnsurePrimaryKey((EntityType)entityType, null, true);
```
Comment (roji) [inline src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs:127]:
```suggestion
            var primaryKeyAttributeExits = CheckAttributesAndEnsurePrimaryKey((EntityType)entityType, null, shouldThrow: true);
```

Comment (roji) [reply src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs:127]: Also: can be inlined (only used once below)

File: src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs [hunk 8]
```diff
@@ -121,4 +162,97 @@ public KeyAttributeConvention(ProviderConventionSetBuilderDependencies dependenc
...
+
+        if (keyBuilder == null)
+        {
+            CheckIgnoredProperties(primaryKeyAttribute, entityType);
+        }
+
+        return true;
+    }
+
+    private static void CheckIgnoredProperties(
```
Comment (roji) [inline src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs:223]: Can be inlined into CheckPrimaryKeyAttributesAndEnsurePrimaryKey

File: src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs [hunk 9]
```diff
@@ -121,4 +162,97 @@ public KeyAttributeConvention(ProviderConventionSetBuilderDependencies dependenc
...
+        PrimaryKeyAttribute primaryKeyAttribute,
+        IConventionEntityType entityType)
+    {
+        foreach (var propertyName in primaryKeyAttribute.PropertyNames)
+        {
+            if (entityType.Builder.IsIgnored(propertyName, fromDataAnnotation: true))
```
Comment (roji) [inline src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs:229]: Not sure, but shouldn't this check be in model validation? i.e. don't we always want this check, no matter which convention sets up a primary key (or even explicit API)?

On a related note, what happens if a primary key property gets ignored later, either by a later convention or explicitly? Just like we support referencing a (shadow) property that doesn't exist yet, maybe we should allow referencing an ignored property that may get un-ignored later?

File: src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs [hunk 10]
```diff
@@ -52,39 +71,49 @@ public KeyAttributeConvention(ProviderConventionSetBuilderDependencies dependenc
...
-            // This code path will also be hit when entity is marked as Keyless explicitly
-            return;
+            return false;
         }
 
-        var entityTypeBuilder = entityType.Builder;
-        var currentKey = entityTypeBuilder.Metadata.FindPrimaryKey();
-        var properties = new List<string> { propertyBuilder.Metadata.Name };
+        var primaryKeyAttributeExists = CheckPrimaryKeyAttributesAndEnsurePrimaryKey(entityType, shouldThrow);
+        var currentKey = entityType.FindPrimaryKey();
```
Comment (roji) [inline src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs:91]: Can put inside inner block below

File: src/EFCore/Properties/CoreStrings.Designer.cs [hunk 11]
```diff
@@ -433,7 +433,7 @@ public static string CompositeFkOnProperty(object? navigation, object? entityTyp
                 navigation, entityType);
 
         /// <summary>
-        ///     The entity type '{entityType}' has multiple properties with the [Key] attribute. Composite primary keys can only be set using 'HasKey' in 'OnModelCreating'.
+        ///     The entity type '{entityType}' has multiple properties with the [Key] attribute. Composite primary keys configured by placing the [PrimaryKey] on the entity type class, or by using 'HasKey' in 'OnModelCreating'.
```
Comment (roji) [inline src/EFCore/Properties/CoreStrings.Designer.cs:436]:
```suggestion
        ///     The entity type '{entityType}' has multiple properties with the [Key] attribute. Composite primary keys configured by placing the [PrimaryKey] attribute on the entity type class, or by using 'HasKey' in 'OnModelCreating'.
```

File: src/EFCore/Properties/CoreStrings.resx [hunk 12]
```diff
@@ -271,7 +271,7 @@
     <value>There are multiple properties with the [ForeignKey] attribute pointing to navigation '{1_entityType}.{0_navigation}'. To define a composite foreign key using data annotations, use the [ForeignKey] attribute on the navigation.</value>
   </data>
   <data name="CompositePKWithDataAnnotation" xml:space="preserve">
-    <value>The entity type '{entityType}' has multiple properties with the [Key] attribute. Composite primary keys can only be set using 'HasKey' in 'OnModelCreating'.</value>
+    <value>The entity type '{entityType}' has multiple properties with the [Key] attribute. Composite primary keys configured by placing the [PrimaryKey] on the entity type class, or by using 'HasKey' in 'OnModelCreating'.</value>
```
Comment (roji) [inline src/EFCore/Properties/CoreStrings.resx:274]:
```suggestion
    <value>The entity type '{entityType}' has multiple properties with the [Key] attribute. Composite primary keys configured by placing the [PrimaryKey] attribute on the entity type class, or by using 'HasKey' in 'OnModelCreating'.</value>
```

File: src/EFCore.Abstractions/PrimaryKeyAttribute.cs [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+///     See <see href="https://aka.ms/efcore-docs-modeling">Modeling entity types and relationships</see> for more information and examples.
+/// </remarks>
+[AttributeUsage(AttributeTargets.Class)]
+public sealed class PrimaryKeyAttribute : Attribute
+{
+    /// <summary>
+    ///     Initializes a new instance of the <see cref="IndexAttribute" /> class.
+    /// </summary>
+    /// <param name="propertyNames">The properties which constitute the index, in order (there must be at least one).</param>
+    public PrimaryKeyAttribute(params string[] propertyNames)
```
Comment (roji) [inline src/EFCore.Abstractions/PrimaryKeyAttribute.cs:24]: Not sure if we've discussed this for IndexAttributer, but should we have a non-params string parameter before the params array? This would force users to specify at least one column in compile-time.

**Review by roji — COMMENTED** [review]
Here's a first round of comments, @AndriySvyryd will probably also want to take a look.

File: src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs (hunk 9 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs:229]: If any properties are ignored then no key would be created for the validation to pick up. And we should be very careful with not looking for attributes outside of conventions for perf reasons.

`NotMappedMemberAttributeConvention` runs before `NotMappedMemberAttributeConvention`.

One of the core properties of the EF Core model building is that at any point (between convention execution) it only contains valid configuration (as far as we know). Making exceptions to this would make it more difficult to work with the unfinished model.

File: src/EFCore/Metadata/Builders/IConventionEntityTypeBuilder.cs [hunk 14]
```diff
@@ -230,6 +230,18 @@ public interface IConventionEntityTypeBuilder : IConventionAnnotatableBuilder
...
+    ///     Sets the properties that make up the primary key for this entity type.
+    /// </summary>
+    /// <param name="propertyNames">The names of the properties that make up the primary key.</param>
+    /// <param name="fromDataAnnotation">Indicates whether the configuration was specified using a data annotation.</param>
+    /// <returns>An object that can be used to configure the primary key.</returns>
+    /// <returns>
+    ///     An object that can be used to configure the primary key if it was set on the entity type,
+    ///     <see langword="null" /> otherwise.
+    /// </returns>
+    IConventionKeyBuilder? PrimaryKey(IReadOnlyList<string> propertyNames, bool fromDataAnnotation = false);
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/IConventionEntityTypeBuilder.cs:243]:
```suggestion
    IConventionKeyBuilder? PrimaryKey(IReadOnlyList<string>? propertyNames, bool fromDataAnnotation = false);
```

File: src/EFCore/Metadata/Conventions/Infrastructure/ProviderConventionSetBuilder.cs [hunk 15]
```diff
@@ -88,6 +90,7 @@ public virtual ConventionSet CreateConventionSet()
         conventionSet.EntityTypeBaseTypeChangedConventions.Add(propertyDiscoveryConvention);
         conventionSet.EntityTypeBaseTypeChangedConventions.Add(servicePropertyDiscoveryConvention);
         conventionSet.EntityTypeBaseTypeChangedConventions.Add(keyDiscoveryConvention);
+        conventionSet.EntityTypeBaseTypeChangedConventions.Add(keyAttributeConvention);
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Conventions/Infrastructure/ProviderConventionSetBuilder.cs:93]: Shouldn't this go before `keyDiscoveryConvention`?

File: test/EFCore.Tests/Metadata/Conventions/PropertyAttributeConventionTest.cs [hunk 16]
```diff
@@ -197,6 +197,16 @@ public void KeyAttribute_does_not_throw_when_setting_composite_primary_key_if_fl
         Assert.Equal("Id", model.FindEntityType(typeof(B)).FindPrimaryKey().Properties[1].Name);
     }
 
+    [ConditionalFact]
+    public void KeyAttribute_does_not_throw_when_setting_composite_primary_key_with_PrimaryKey_attribute()
```
Comment (AndriySvyryd) [inline test/EFCore.Tests/Metadata/Conventions/PropertyAttributeConventionTest.cs:201]: Also add a test where a [PK] is on base type while the derived type has [Key] (it shouldn't throw)

File: src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs (hunk 9 above)
Comment (ajcvickers) [reply src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs:229]: Copied this from IndexAttribute. Will change in both places.

File: src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs (hunk 3 above)
Comment (ajcvickers) [reply src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs:65]: This isn't something I changed. I think it was done like this because we need to know the property that has the attribute, but we currently don't store that information in the final model--because it doesn't form part of the model. So I think if we want to do it at other times, then we would have to discover (or re-discover) any Key attributes to know that we should warn. If we think it's worth doing this, then I'll file an issue for it, since I think it should be in separate PR.

File: src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs (hunk 7 above)
Comment (ajcvickers) [reply src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs:127]: It has side-effects; specifically, throwing on error cases.

**Comment by ajcvickers** [comment]
New version up.

File: src/EFCore.Abstractions/PrimaryKeyAttribute.cs [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace Microsoft.EntityFrameworkCore;
+
+/// <summary>
+///     Specifies a primary key for the entity type mapped to this CLR type.
+/// </summary>
+/// <remarks>
+///     <para>
+///         This attribute can be used for both keys made up of a
+///         single property, and for composite keys made up of multiple properties. `System.ComponentModel.DataAnnotations.KeyAttribute`
```
Comment (AndriySvyryd) [inline src/EFCore.Abstractions/PrimaryKeyAttribute.cs:14]:
```suggestion
///         single property, and for composite keys made up of multiple properties. <see cref="KeyAttribute" />
```

Comment (AndriySvyryd) [inline src/EFCore.Abstractions/PrimaryKeyAttribute.cs:14]:
```suggestion
///         single property, and for composite keys made up of multiple properties. <see cref="KeyAttribute" />
```

File: src/EFCore/Metadata/Conventions/IndexAttributeConvention.cs [hunk 18]
```diff
@@ -119,67 +165,9 @@ public IndexAttributeConvention(ProviderConventionSetBuilderDependencies depende
 
                 if (indexBuilder is not null && indexAttribute.IsDescending is not null)
                 {
-                    indexBuilder = indexBuilder.IsDescending(indexAttribute.IsDescending, fromDataAnnotation: true);
+                    indexBuilder.IsDescending(indexAttribute.IsDescending, fromDataAnnotation: true);
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Conventions/IndexAttributeConvention.cs:168]: Why was this changed?

Comment (AndriySvyryd) [inline src/EFCore/Metadata/Conventions/IndexAttributeConvention.cs:168]: Why was this changed?

File: src/EFCore/Metadata/Conventions/IndexAttributeConvention.cs [hunk 19]
```diff
@@ -100,15 +100,61 @@ public IndexAttributeConvention(ProviderConventionSetBuilderDependencies depende
                 }
                 catch (InvalidOperationException exception)
                 {
-                    CheckMissingProperties(indexAttribute, entityType, exception);
+                    foreach (var propertyName in indexAttribute.PropertyNames)
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Conventions/IndexAttributeConvention.cs:103]: Methods should be extracted if it helps clarity.

Comment (AndriySvyryd) [inline src/EFCore/Metadata/Conventions/IndexAttributeConvention.cs:103]: Methods should be extracted if it helps clarity.

File: src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs [hunk 20]
```diff
@@ -95,30 +124,124 @@ public KeyAttributeConvention(ProviderConventionSetBuilderDependencies dependenc
...
+        IConventionEntityType entityType,
+        bool shouldThrow)
+    {
+        var primaryKeyAttribute = entityType.ClrType.GetCustomAttributes<PrimaryKeyAttribute>(inherit: true).FirstOrDefault();
+        if (primaryKeyAttribute == null)
+        {
+            return false;
+        }
+
+        if (entityType.ClrType.GetCustomAttributes<KeylessAttribute>(inherit: true).Any())
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Conventions/KeyAttributeConvention.cs:179]: Use `Attribute.IsDefined`

**Review by AndriySvyryd — APPROVED** [review]

File: src/EFCore/Metadata/Conventions/IndexAttributeConvention.cs (hunk 19 above)
Comment (ajcvickers) [reply src/EFCore/Metadata/Conventions/IndexAttributeConvention.cs:103]: Same methods exist in both conventions. @roji suggested inlining. I don't mind one way other the other.

**Comment by paulodero** [comment]
Great feature! A much better way to define composite keys in `efcore`.
